dora/core/server/master/src/main/java/alluxio/master/scheduler/Scheduler.java

```diff
@@ -558,6 +393,163 @@ public JobMetaStore getJobMetaStore() {
+      justification = "Already performed null check")
+  public class WorkerInfoHub {
+    public Map<WorkerInfo, CloseableResource<BlockWorkerClient>> mActiveWorkers = ImmutableMap.of();
+    private final WorkerProvider mWorkerProvider;
+
+    /**
+     * Constructor.
+     * @param scheduler scheduler
+     * @param workerProvider worker provider
+     */
+    public WorkerInfoHub(Scheduler scheduler, WorkerProvider workerProvider) {
+      mWorkerProvider = workerProvider;
+    }
+
+    private final Map<WorkerInfo, PriorityBlockingQueue<Task>> mWorkerToTaskQ
+        = new ConcurrentHashMap<>();
+
+    /**
+     * Enqueue task for worker.
+     * @param workerInfo the worker
+     * @param task the task
+     * @param kickStartTask kick-start task
+     * @return whether the task is enqueued successfully
+     */
+    public boolean enqueueTaskForWorker(@Nullable WorkerInfo workerInfo, Task task,
+        boolean kickStartTask) {
+      if (workerInfo == null) {
+        return false;
+      }
+      PriorityBlockingQueue workerTaskQ = mWorkerToTaskQ
+          .computeIfAbsent(workerInfo, k -> new PriorityBlockingQueue<>());
+      if (!workerTaskQ.offer(task)) {
+        return false;
+      }
+      CloseableResource<BlockWorkerClient> blkWorkerClientResource = mActiveWorkers.get(workerInfo);
+      if (blkWorkerClientResource == null) {
+        LOG.warn("Didn't find corresponding BlockWorkerClient for workerInfo:{}",
+            workerInfo);
+        return false;
+      }
+      if (kickStartTask) {
+        // track running tasks of a job
+        ConcurrentHashSet<Task<?>> tasks = mJobToRunningTasks.computeIfAbsent(task.getJob(),
+                j -> new ConcurrentHashSet<>());
+        tasks.add(task);
+        task.execute(blkWorkerClientResource.get(), workerInfo);
+        task.getResponseFuture().addListener(() -> {
+          Job job = task.getJob();
+          try {
+            job.processResponse(task); // retry on failure logic inside
+            // TODO(lucy) currently processJob is only called in the single
+            // threaded scheduler thread context, in future once tasks are
+            // completed, they should be able to call processJob to resume
+            // their own job to schedule next set of tasks to run.
+          } catch (Exception e) {
+            // Unknown exception. This should not happen, but if it happens we don't
+            // want to lose the worker thread, thus catching it here. Any exception
+            // surfaced here should be properly handled.
+            LOG.error("Unexpected exception thrown in response future listener.", e);
+            job.failJob(new InternalRuntimeException(e));
+          } finally {
+            // whether task succeed or fail, remove it from q,
+            workerTaskQ.remove(task);
+            ConcurrentHashSet<Task<?>> runningTaskSet =
+                mJobToRunningTasks.getOrDefault(job, new ConcurrentHashSet<>());
```

Comment:
change to getOrDefault since we probably would delete the job in mJobToRunningTasks when we still have leftover tasks

Reply:
then u can still use compute

Reply:
addressed